### PR TITLE
add systemd-resolved when `boot.initrd.systemd.network` is enabled

### PR DESCRIPTION
**Pull request by mlyxshi**
## Description of changes

Upstream systemd-resolved  part  in https://github.com/ElvishJerricco/stage1-tpm-tailscale.
@ElvishJerricco

#301333

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by r-vdp** [comment]
I think you missed the line that includes the resolved config file in the initrd as well, so that we get the same config as what we have in stage 2.

**Comment by mlyxshi** [comment]
> I think you missed the line that includes the resolved config file in the initrd as well, so that we get the same config as what we have in stage 2.

I actually don't know. I just think the stage 2 related stuff shouldn't be included in stage1. 

https://github.com/NixOS/nixpkgs/blob/219951b495fc2eac67b1456824cc1ec1fd2ee659/nixos/modules/system/boot/systemd/initrd.nix#L387

**Comment by mlyxshi** [comment]
> I think you missed the line that includes the resolved config file in the initrd as well, so that we get the same config as what we have in stage 2.
> 
> https://github.com/ElvishJerricco/stage1-tpm-tailscale/blob/main/initrd-tailscale.nix#L37

For personal project, defining stage1 resolved config file as same as stage 2 is convenient.
However, as you can see in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/networkd.nix#L2990, in initrd `/etc/systemd/networkd.conf` is defined by `boot.initrd.systemd.network.config` not stage 2 config.

Actually, the default `systemd-resolved` configuration is enough for most cases, people really care about resolved configuration should config it via `boot.initrd.systemd.contents"/etc/systemd/resolved.conf".text = ""`
I think we need to implement something like `boot.initrd.systemd.resolved.config` in another pr.

File: nixos/modules/system/boot/networkd.nix [hunk 1]
```diff
@@ -2995,13 +2998,18 @@ let
       systemd.sockets.systemd-networkd = {
         wantedBy = [ "initrd.target" ];
       };
+      systemd.services.systemd-resolved = {
+        wantedBy = ["initrd.target"];
+        serviceConfig.ExecStartPre = "-+/bin/ln -s /run/systemd/resolve/resolv.conf /etc/resolv.conf";
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/networkd.nix:3003]: I think this should be a tmpfiles rule, really.

File: nixos/modules/system/boot/networkd.nix [hunk 2]
```diff
@@ -2995,13 +2998,18 @@ let
       systemd.sockets.systemd-networkd = {
         wantedBy = [ "initrd.target" ];
       };
+      systemd.services.systemd-resolved = {
+        wantedBy = ["initrd.target"];
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/networkd.nix:3002]: The upstream unit file has `WantedBy=sysinit.target`, so we should probably do the same. Same goes for the stage 2 unit, really.

**Review by ElvishJerricco — CHANGES_REQUESTED** [review]
I think this should all go in `nixos/modules/system/boot/resolved.nix` rather than `networkd.nix`

File: nixos/modules/system/boot/networkd.nix (hunk 1 above)
Comment (mlyxshi) [reply nixos/modules/system/boot/networkd.nix:3003]: I have no idea how to do it properly. It would be better for you to upstream this

**Pull request closed without merging**